### PR TITLE
feat: Add GitLab forge Integration to Git-Spice

### LINKED ISSUE #348: Feature Request: Gitlab Support
It would be nice if git-spice also supported creating Merge Requests in GitLab

If it's not supported, it would also already be helpful to allow pushing all branches in the stack after restacking

### PR DESCRIPTION
**Pull request by gr4cza**
This pull request introduces support for GitLab integration, including authentication, metadata handling, comment management, change editing, and finding changes. 

For authentication, it only implements _Personal Access Token_ auth, because I wasn't able to make the [Device authorization grant flow](https://docs.gitlab.com/ee/api/oauth2.html#device-authorization-grant-flow) work on [gitlab.com](gitlab.com).

For the GitLab API integration, it uses [xanzy/go-gitlab](https://github.com/xanzy/go-gitlab) SDK.

There is a lot of duplicated code in the implementation since there are many similarities between the two forges, but I did not want to mess with the already existing code at first.

What missing:
- [ ]  Doc update
- [x]  Test copy + exstension

Feel free to modify this as needed to better fit the specifics of your work.

Resolves #348

### CODE REVIEW

**Comment by abhinav** [comment]
Cool! I'm okay with starting with PAT-only and expanding to OAuth, apps, and/or GitLab CLI in the future.

I skimmed over the PR and high-level looks great.
I'll try to go over it in detail more by this weekend, but a couple minor fixes until then:

-  The only failing integration test is the one where it prompts for the forge and selects one.

    ```
            ### initial ###
            Select a Forge:
            
            ▶ github
              gitlab
              shamhub
            ### select ###
            Select a Forge:
            
              github
            ▶ gitlab
              shamhub
    ```

    That's an easy fix:

    https://github.com/abhinav/git-spice/blob/b6cf1636883c30500ecbe9d1d3971846a3b94029/testdata/script/auth_prompt_forge.txt#L18

    If you turn that into `\x1b[B\x1b[B` (down twice), that'll fix that test.

- Please run [gofumpt](https://github.com/mvdan/gofumpt) over the code to get formatting fixed.
- The PAT auth option description still refers to GitHub. You probably meant to change that to GitLab.

Have you had a chance to try a build of this on your GitLab repositories?
I'm curious how well it worked. I've dusted off my old GitLab account to experiment with it when I have a second as well.

Again, thanks for putting this together. I'm happy to collaborate on the tests and docs needed to get this merged.
I'll follow up on the documentation website separately after this is merged—the website assumes a lot of GitHub and "PR" in its instructions, and restructuring that isn't in-scope for this PR.

**Comment by gr4cza** [comment]
Thank you for your quick reply!

There are two major differences in the implementation. I will detail them below.

I also have a couple of questions:
- I saw that you used URLs for `ChangeComments` here:
  https://github.com/gr4cza/git-spice/blob/c8f793f96ed0fedaf0a6d571cdad33189168f73d/internal/forge/gitlab/comment.go#L31-L33
  and here:
  https://github.com/gr4cza/git-spice/blob/c8f793f96ed0fedaf0a6d571cdad33189168f73d/internal/forge/gitlab/comment.go#L51
  Is it necessary because the GitLab API does not return any link to a comment (note), so we can only calculate it on the fly?

- How did you generate/capture the fixture files under `internal/forge/github/testdata/fixtures`? If there is an easy method, I would like to use that. :D

I'm new to Go (since I'm mainly a Kotlin dev) and was not familiar with the tooling at all, which is why I did not run `gofumpt` yet. How common is the knowledge that it needs to be used? If it’s not, you may want to add it to the contribution guide as well (along with `golangci-lint`).

Are you planning to squash the commits? If yes, I won’t bother with keeping the commit history clean during the fixes.

I developed it while constantly testing it locally with gitlab.com and our private forge as well, and it seems to work fine.

File: internal/forge/gitlab/comment.go
Comment (gr4cza) [inline internal/forge/gitlab/comment.go:107]: The GitLab API does not provide information about whether you can update a comment or not, so I am calculating it based on this [Edit a comment](https://docs.gitlab.com/ee/user/discussions/#edit-a-comment).
> You can edit your own comment at any time. Anyone with at least the Maintainer role can also edit a comment made by someone else.

File: internal/forge/gitlab/edit.go
Comment (gr4cza) [inline internal/forge/gitlab/edit.go:26]: GitLab determines the MR readiness based on the title, so we have to fetch the current title and modify it accordingly.  [Mark merge requests as drafts](https://docs.gitlab.com/ee/user/project/merge_requests/drafts.html#mark-merge-requests-as-drafts)
>Creating or editing a merge request: Add [Draft], Draft: or (Draft) to the beginning of the merge request’s title, or select Mark as draft below the Title field.

>Editing an existing merge request: Remove [Draft], Draft: or (Draft) from the beginning of the title, or clear Mark as draft below the Title field.

**Comment by abhinav** [comment]
> I saw that you used URLs for ChangeComments here [..]
> Is it necessary because the GitLab API does not return any link to a comment (note), so we can only calculate it on the fly?

It's not necessary at all. They didn't end up getting used at all, e.g. we don't post them to the user.

> How did you generate/capture the fixture files under internal/forge/github/testdata/fixtures?

The setup for that is in github/integration_test.go:

- Add an `-update` flag to the test binary:

    https://github.com/abhinav/git-spice/blob/b6cf1636883c30500ecbe9d1d3971846a3b94029/internal/forge/github/integration_test.go#L37

- A recorder that uses that flag to determine whether to record requests/responses or replay.

    https://github.com/abhinav/git-spice/blob/b6cf1636883c30500ecbe9d1d3971846a3b94029/internal/forge/github/integration_test.go#L48-L54

    This is based on the gopkg.in/dnaeon/go-vcr.v4 library. It supports doing record/replay for HTTP requests.

- Create a GitHub client based on the recorder and feed it into the newRepository:

    https://github.com/abhinav/git-spice/blob/b6cf1636883c30500ecbe9d1d3971846a3b94029/internal/forge/github/integration_test.go#L146-L149

    Note that NewRepository is not exposed from the public API of the package.
    It's made visible just to tests in that directory with this:

    https://github.com/abhinav/git-spice/blob/b6cf1636883c30500ecbe9d1d3971846a3b94029/internal/forge/github/repository_int_test.go#L1-L5

    (Normally, tests in that directory would be able to access newRepository anyway because they reside in the same package, but the `package github_test` on top of the integration test puts it in a separate Go package that must import it explicitly.)

With the above machinery in place, I wrote reproducible integration tests that exercise the various forge functions.
I ran them with `-update` as I wrote them to get the fixtures.

For tests that needed random input (e.g. a random branch name or a random text file name), there was also a different need: random input had to be random at record time (the first time the test is run), but had to remain deterministic at replay time.
That's provided by my [fixturetest package](https://github.com/abhinav/git-spice/blob/b6cf1636883c30500ecbe9d1d3971846a3b94029/internal/fixturetest/fixturetest.go):

- Set up at the top of the integration test:

    https://github.com/abhinav/git-spice/blob/b6cf1636883c30500ecbe9d1d3971846a3b94029/internal/forge/github/integration_test.go#L38

- Generate a random string the first time around, but reuse it after that:

    https://github.com/abhinav/git-spice/blob/b6cf1636883c30500ecbe9d1d3971846a3b94029/internal/forge/github/integration_test.go#L286-L290

    (Feel free to move the [randomString function](https://github.com/abhinav/git-spice/blob/b6cf1636883c30500ecbe9d1d3971846a3b94029/internal/forge/github/integration_test.go#L583-L593) into the fixturetest package to re-use if you need it.)

> I did not run gofumpt yet. How common is the knowledge that it needs to be used? If it’s not, you may want to add it to the contribution guide as well (along with golangci-lint).

Yeah, that makes sense. It's common knowledge to use gofmt. (which comes with Go), but gofumpt, which is a more opinionated third-party formatter is not "expected". I'll follow up on the contribution guide changes.

> Are you planning to squash the commits? If yes, I won’t bother with keeping the commit history clean during the fixes.

Yep! I will do a merge commit only if you have a carefully crafted commit history, so feel free to use whichever path you find easier to work with and I'll squash or merge accordingly. Personally I tend to go the squash route for most of my work.

> I developed it while constantly testing it locally with gitlab.com and our private forge

That's pretty great. Thanks!

**Comment by abhinav** [comment]
@gr4cza Just a heads up: I've started pushing to your branch, so I suggest avoiding `push -f` while we're both contributing to this branch to avoid stomping on changes.

**Comment by gr4cza** [comment]
@abhinav thanks for the heads up, and your fixes! I'm currently working on the integration tests. My plan is to finish them during the weekend.

**Comment by abhinav** [comment]
Oh excellent, I'm refactoring the GitHub integration test so that some of the infrastructure is easier to use in the GitLab integration test. I'll avoid adding GitLab integration tests for now.

**Comment by abhinav** [comment]
@gr4cza I've taken a full pass over the PR. I tweaked a few things here and there, and refactored a couple minor things. Overall this is looking great!

The integration tests may need access to a test repository similar to [this one on GitHub](https://github.com/abhinav/test-repo). I've set one up here: https://gitlab.com/abg/test-repo. If you share your GitLab username with me, I can give you access to it so you can run the initial recording of the requests made by the integration tests.

**Comment by gr4cza** [comment]
@abhinav  I'm running under the same @gr4cza nick in GitLab.

**Comment by abhinav** [comment]
Invited. Let me know if the push access isn't enough and I'll bump it.

**Comment by abhinav** [comment]
@gr4cza I got OAuth2 working so I've added that as an auth option.

File: internal/forge/gitlab/client.go
Comment (abhinav) [inline internal/forge/gitlab/client.go:23]: this is the important bit: gitlab library uses a different client based on the token type, so we need to record what auth method was used: pat or oauth.

File: internal/forge/gitlab/auth.go
Comment (gr4cza) [inline internal/forge/gitlab/auth.go:209]: do we really need all three scopes here?  [GitLab Doc](https://docs.gitlab.com/ee/integration/oauth_provider.html#view-all-authorized-applications)

api | Grants complete read/write access to the API, including all groups and projects, the container registry, the dependency proxy, and the package registry.
-- | --

Comment (abhinav) [reply internal/forge/gitlab/auth.go:209]: (Long term, r/w on all api is quite broad. I'm hoping there's a narrower set I can put in place here.)

**Comment by abhinav** [comment]
@gr4cza I forgot to mention this: If you have any trouble with the integration test, please feel free to push what you have an we can iterate on it together. 

Also, a heads up: I'll be merging #480 shortly which should have some conflicts with this PR. I will push an update to your branch once that merges to resolves those conflicts.

**Comment by gr4cza** [comment]
@abhinav I think I was able to resolve most of the initial problems (differences from the GitHub API) and now I'm copying the test cases. Unfortunately, I'm not allowed to push (and I think I will not be able to merge either), so please give me more elevated rights in the test repo over there.

**Comment by abhinav** [comment]
Done!

**Comment by gr4cza** [comment]
@abhinav I've added the integration tests :D

**Review by abhinav — APPROVED** [review]
This is great!
I've got a local branch updating the doc website, but I think this PR is good to merge as-is.

Thanks so much for putting the effort into this!

**Pull request merged**
